logger/logger.go

```diff
@@ -22,45 +24,57 @@ func (tfLogger *TfLogger) Enabled(_ context.Context, _ logger.Level) bool {
 
 func (tfLogger *TfLogger) Tracef(ctx context.Context, format string, v ...any) {
 	if tfLogger == nil {
-		tflog.Trace(ctx, fmt.Sprintf(format, v...), nil)
+		tflog.Trace(getLoggerValidContext(ctx), fmt.Sprintf(format, v...), nil)
 	} else {
-		tflog.SubsystemTrace(ctx, tfLogger.Name, fmt.Sprintf(format, v...), nil)
+		tflog.SubsystemTrace(getLoggerValidContext(ctx), tfLogger.Name, fmt.Sprintf(format, v...), nil)
 	}
 }
 
 func (tfLogger *TfLogger) Debugf(ctx context.Context, format string, v ...any) {
 	if tfLogger == nil {
-		tflog.Debug(ctx, fmt.Sprintf(format, v...), nil)
+		tflog.Debug(getLoggerValidContext(ctx), fmt.Sprintf(format, v...), nil)
 	} else {
-		tflog.SubsystemDebug(ctx, tfLogger.Name, fmt.Sprintf(format, v...), nil)
+		tflog.SubsystemDebug(getLoggerValidContext(ctx), tfLogger.Name, fmt.Sprintf(format, v...), nil)
 	}
 }
 
 func (tfLogger *TfLogger) Infof(ctx context.Context, format string, v ...any) {
 	if tfLogger == nil {
-		tflog.Info(ctx, fmt.Sprintf(format, v...), nil)
+		tflog.Info(getLoggerValidContext(ctx), fmt.Sprintf(format, v...), nil)
 	} else {
-		tflog.SubsystemInfo(ctx, tfLogger.Name, fmt.Sprintf(format, v...), nil)
+		tflog.SubsystemInfo(getLoggerValidContext(ctx), tfLogger.Name, fmt.Sprintf(format, v...), nil)
 	}
 }
 
 func (tfLogger *TfLogger) Warnf(ctx context.Context, format string, v ...any) {
 	if tfLogger == nil {
-		tflog.Warn(ctx, fmt.Sprintf(format, v...), nil)
+		tflog.Warn(getLoggerValidContext(ctx), fmt.Sprintf(format, v...), nil)
 	} else {
-		tflog.SubsystemWarn(ctx, tfLogger.Name, fmt.Sprintf(format, v...), nil)
+		tflog.SubsystemWarn(getLoggerValidContext(ctx), tfLogger.Name, fmt.Sprintf(format, v...), nil)
 	}
 }
 
 func (tfLogger *TfLogger) Errorf(ctx context.Context, format string, v ...any) {
 	if tfLogger == nil {
-		tflog.Error(ctx, fmt.Sprintf(format, v...), nil)
+		tflog.Error(getLoggerValidContext(ctx), fmt.Sprintf(format, v...), nil)
 	} else {
-		tflog.SubsystemError(ctx, tfLogger.Name, fmt.Sprintf(format, v...), nil)
+		tflog.SubsystemError(getLoggerValidContext(ctx), tfLogger.Name, fmt.Sprintf(format, v...), nil)
+	}
+}
+
+// Tflogger is called through Databricks provider and also Go SDK
+// We need a way to check if the context is correctly set for logging, if not then we use the context we define in configureContextFunc
+// We can't use the logging.GetProviderRootLogger(ctx) to check as it is internal
+func getLoggerValidContext(ctx context.Context) context.Context {
+	logger := ctx.Value("provider")
+	if logger == nil {
+		return loggerContext
 	}
+	return ctx
 }
 
-func SetLogger() {
+func SetLogger(ctx context.Context) {
```

Comment:
I'm a little puzzled by this function (especially the previous version) as it is resetting something. If I understand correctly, it could be rewritten as follows:

```go
func SetLogger(ctx context.Context) {
	logger.DefaultLogger = nil
	loggerContext = ctx
}
```

Is that the intended behavior? If so, I think it would benefit from a comment to explain why we are resetting the defaultLogger.

Reply:
`DefaultLogger` is defined in Go SDK. We set it only at this place inside terraform repo (we also set this to SimpleLogger in GoSDK). We can't re-write this to nil as we need to specify the type to `TfLogger`. 

So when this is run inside Go SDK (as part of terraform provider), we get the logger based on DefaultLogger: https://github.com/databricks/databricks-sdk-go/blob/main/logger/logger.go#L47

Reply:
Both SimpleLogger (defined in Go SDK) and TfLogger (defined in TF repo) implements the Logger (defined in Go SDK) interface

Reply:
OK, I see that a `nil` TfLogger is actually a valid logger.